### PR TITLE
Add higher-level support for "bundles" and "collections" of resources

### PR DESCRIPTION
**Pull request by totten**
Overview
----------------------------------------

Every HTML page is built on a list of "resources" (aka "assets") such as Javascript and CSS files. Civi provides a few inter-related APIs for managing these (`CRM_Core_Resources`, `CRM_Core_Region`). This branch expands on these in a few ways:

1. Add support for "Resource Bundles" (`CRM_Core_Resources_Bundle`). By using a bundle, an application-page may activate several resources with a single step.
   ```php
   // Before
   Civi::resources()->addScriptFile('civicrm', 'js/foobar.js')
     ->addStyleFile('civicrm', 'css/foobar.css')
     ->addVars('foobar', computeTheFooBar());

   // After
   Civi::resources()->addBundle('foobar');
   ```
   Observe that we give a symbolic name (`foobar`) to the list of related resources (`js/foobar.js`, `css/foobar.css`, and `CRM.vars.foobar`).
2. Retrofit the "bundle" concept into a few pre-existing dataflows (`CRM_Core_Resources::addCoreResources()`, `CRM_Core_Resources::addCoreStyles()`).
3. Add an abstract concept, "Resource Collection". This embodies the common elements shared by `CRM_Core_Resources`, `CRM_Core_Resources_Bundle`, and `CRM_Core_Region`. To wit: the methods `add()`, `addScriptFile()`, `addScriptUrl()`, etc. are supported across the board. They are designated via PHP interfaces/traits.

The resulting model looks like so:

Screenshot from 2020-08-18 16-14-02

(*Note: This contributes toward https://lab.civicrm.org/dev/user-interface/-/issues/27.*)

Before
----------------------------------------

(1) There is no pattern for giving a name to a "mixable list of resources".

(2) There are three places where you might add, inspect, or tap-into a list of resources. These three build on top of each other, and they are broadly similar, but there are technical differences. 

| Criterion       | `CRM_Core_Resources`      | `hook_coreResourcesList`       | `CRM_Core_Region` |
| -- | -- | -- | -- |
| Method style | `addFoo('bar')` | `$arr[] = 'bar'` | `add(['foo' =>'bar'])` |
| IDE Discovery   | Great                   | Poor                    | OK |
| Alterability    | Poor                    | Great                   | OK | 
| Ordering | By `weight` | By array position | By `weight` |
| Singleton/Lego     | Singleton-like          | Singleton-strict        | Lego-like |
| Content ID  | None                    | URL            | Name (if given) or URL/Number |
| Content type: ScriptFile      | Yes                     | No                      | No |
| Content type: ScriptURL       | Yes                     | Yes                     | Yes |
| Content type: ScriptCode      | Yes                     | No                      | Yes |
| Content type: StyleFile       | Yes                     | No                      | No |
| Content type: StyleURL        | Yes                     | Yes                     | Yes |
| Content type: StyleCode       | Yes                     | No                      | Yes |
| Content type: Setting/Var     | Yes                     | Yes                     | No |
| Content type: Strings         | Yes                     | No                      | No |

Some notable differences between them:

* `addScriptUrl()`, `addScriptFile()`, etc are in `CRM_Core_Resources` but not in `CRM_Core_Region` or `hook_coreResourcesList`
* `add()` is in `CRM_Core_Region` but not in `CRM_Core_Resources` or `hook_coreResourcesList`
* `hook_coreResourceList` allows extensions to add/filter/remove resources. However, it only supports a couple resource-types, and strictness of URLs not always jive with dynamic assets.

After
----------------------------------------

(1) `CRM_Core_Resources_Bundle` defines a "list of resources" that is *available* for use -- but is not necessarily activated on every page-load.

(2) The different collection-like APIs (`CRM_Core_Resources`, `CRM_Core_Region`, etc) all implement the same PHP interfaces (`CollectionInterface` and `CollectionAdderInterface`) to ensure similar usage/behavior.

|   | `CollectionInterface` / `CollectionTrait`? | `CollectionAdderInterface` / `CollectionAdderTrait`? |
| -- | -- | -- |
| __Methods__ | `get()`, `getAll()`, `find()`, `filter()`, `update()` | `add()`, `addScriptFile()`, `addStyleFile()`, `addSetting()`, ad nauseum  |
| __Implementations__ |  |  |
| `CRM_Core_Region` | Yes | Yes |
| `CRM_Core_Resources_Bundle` | Yes | Yes |
| `CRM_Core_Resources` | No | Yes |

(3) Every instance of `CRM_Core_Resources_Bundle` can be altered via `hook_civicrm_alterBundle($bundle)`.

* This is a replacement for `hook_coreResourcesList`. It works with more bundles, allows more content-types, and provides helper methods like `filter()`, `find()`, and `clear()`.
* For backward-compatibility, we still fire `hook_coreResourcesList`.

(4) There are two built-in bundles, `coreStyles` and `coreResources`. These are defined in `CRM_Core_Resources_Common`.

### CODE REVIEW

**Comment by totten** [comment]
The patch is fairly long, so it may be hard to trace the refactorings point-by-point. As an alternative QA, I would suggest these as possible tasks:

* Browse the code on the branch -- see how `CollectionTrait` and `CollectionInterface` are used. Think about whether they look compatible.
* Browse the unit-tests. (In particular, check `CollectionTraitTest` and the list of examples which are supposed to be equivalent.)
* Think of a few extensions or use-cases which lean on `CRM_Core_Resources` or `CRM_Core_Region`. For each one, grab a copy of the HTML before patching -- and compare it to the HTML after patching.
    * Example -- here's what did:
        * I picked ~8 pages and wrote a [quickie QA script](https://gist.github.com/totten/b33d89e37e69cd9efe5215865fc8498e) to grab the content of each. I did this on a plain `dmaster` -- and with the combination of `shoreditch`+`org.civicrm.civicase`.
        * Then I ran `colordiff -ru OLD_SNAPSHOTS NEW_SNAPSHOTS` to compare the snapshots (before/after patching). This helped me see changes during development.
        * On the current branch, there is only one change that shows up in the diff. Within the `page-footer`, the location of `<link href="http://dmaster.bknix:8001/sites/all/modules/civicrm/css/admin.css?r=Nw8gU" rel="stylesheet" type="text/css"/>` has changed. I believe the reason is that `styleFile`s are now a formal resource-type, and the default `name` is slightly different:
            * __Before__: Resources are sorted on `weight,name`. The effective name of this resource was its *full URL* (`http://dmaster.bknix.../admin.css?r=Nw8gU`). This name puts it after the `default` content.
            * __After__: Resources are still sorted on `weight,name` The effective name of this resource is now the ext-key and file-id (`civicrm:css/admin.css`). This name puts it before the `default` content.
            * __Comment__:  I'm reallly struggling to care about this one. I guess if we could change the naming formula...
    * It would be good for a reviewer to try a similar exercise - but with other use-cases. Maybe y'all can think of more than me... but a few that come to my mind:
        * Enable another theme/extension (not shoreditch). Grab a few frontend and backend pages.
        * Enable a payment-processor that requires JS assets. Grab the online contribution screens for frontend and backend users.
        * Enable an extension (like Search Builder) which relies on loadPage/loadForm. Grab a few of those AJAX requests.

**Comment by mattwire** [comment]
@totten Would you be able to review/merge https://github.com/civicrm/civicrm-core/pull/18141 before I go too deep into this PR. That "fixes" an edge-case with addvars that causes problems for me with Stripe.

I think I understand the need for something like this PR - given that I now have a fairly complex set of requirements and optional requirements for Stripe which together make for the best user experience (and in some situations make it work). I'm increasing running into issues around load-order and tracking that. Perhaps we could have a chat to discuss a bit further?

**Comment by totten** [comment]
@mattwire - OK, cool. I'll try to take a read on #18141 after I get up tomorrow. I hadn't seen the PR before. At first glance, it looks like the gist of it is to store settings on a per-region basis. I'm rather glad you've come to that 😌 ... that was a characteristic I needed along the way, but I wasn't sure if would make sense from other POV.  

If I understand correctly, it's roughly the same goal as 9a651886d47eb8cf4e07381577d754ead86084fb, with a difference in where that storage specifically lives:

* In 99c6336fda08f705fc2df0ceb2081efd3034ae4a, the settings move into `$resources->settings[$region]`
* In 9a651886d47eb8cf4e07381577d754ead86084fb, the settings move into `$region->_snippets['settings']`.

But I probably need to read 99c6336fda08f705fc2df0ceb2081efd3034ae4a more carefully to grok the motivation, the edges (like initialization/emptiness/etc), and the tests.

(FWIW, the motivation in 9a651886d47eb8cf4e07381577d754ead86084fb is to allow logic like `$pageFooterRegion->merge($bootstrapBundle->getAll())`. This requires that all valid resource-types -- scripts, styles, settings, etc -- be addressable within the `$pageFooterRegion` and the `$bootstrapBundle`.)

**Comment by eileenmcnaughton** [comment]
@colemanw I feel like you know this code the best

File: CRM/Core/Region.php [hunk 1]
```diff
@@ -131,7 +131,7 @@ public function render($default, $allowCmsOverride = TRUE) {
         case 'settings':
           $settingsData = $this->getSettings();
 
-          if (($this->_name === 'html-header') || !CRM_Core_Resources::isAjaxMode()) {
+          if (($this->_name === 'html-header') || (($this->_name !== 'billing-block') && !CRM_Core_Resources::isAjaxMode())) {
```
Comment (totten) [inline CRM/Core/Region.php:134]: @mattwire I took this change from #18141 along with the unit-test revisions and forward-ported via d279541befb1fe74df614fab2d6549984223430d. I'm not sure the process to reproduce/validate via `r-run`, but the unit test passes.

Tangentially... it's

Comment (eileenmcnaughton) [reply CRM/Core/Region.php:134]: The bug it's supposed to address is when you have more then 1 processor on a page & you switch js doesn't load

Matt had some patches against Omnipay which I merged but then this was logged 
https://github.com/eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor/issues/165 and I reverted. I have not looked into it yet but in conjunction with nz.co.fuzion.omnipaymultiprocessor it should work).

So, currently I believe Paypal checkout should load if it is the default processor (but js errors if not). Without the revert the reverse would be true & with the new patch + this it should hopefully work (I have creds if you actually want to try)

Comment (totten) [reply CRM/Core/Region.php:134]: Thanks @eileenmcnaughton. So in theory I could use that to QA the original #18141 and/or my attempted forward-port (d279541befb1fe74df614fab2d6549984223430d). I tried and can't reproduce what's supposed to be wrong... My report-back is a bit long/detailed, and it's not really about bundles/collections, so I posted on the other PR: https://github.com/civicrm/civicrm-core/pull/18141#issuecomment-679916939

Comment (totten) [reply CRM/Core/Region.php:134]: Just a bit of follow-up to round-off this subthread - the discussion in 18141 did give a testing procedure, and this forward-port does appear to be working.

**Comment by eileenmcnaughton** [comment]
I'm generally OK with this PR & at this stage

File: CRM/Core/Region.php (hunk 1 above)
Comment (colemanw) [reply CRM/Core/Region.php:134]: I've got an alternate proposal about how to render settings here: #18262

**Comment by eileenmcnaughton** [comment]
We discussed this on a CT call this morning and the consensus was now that the rc is cut we should merge this and really try to focus in on the greenwich-as-theme that it lays the groundwork for. It will have 2 full months between merge & release

File: Civi/Core/Container.php [hunk 2]
```diff
@@ -206,6 +206,12 @@ public function createContainer() {
       []
     ))->setPublic(TRUE);
 
+    $container->setDefinition('bundle.coreStyles', new Definition('CRM_Core_Resources_Bundle', ['coreStyles']))
+      ->setFactory('CRM_Core_Resources_Common::createStyleBundle');
+
```
Comment (demeritcowboy) [inline Civi/Core/Container.php:211]: Do these need setPublic(TRUE) or is there something you need to configure on a site? I'm having trouble after updating a drupal 8 site and adding setPublic makes it work.

**Pull request merged**